### PR TITLE
i#2985 drx_expand_scatter_gather(): Add drx restore state machine for AVX2 gather.

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds support for restoring the destination mask in the AVX2 gather emulation sequence. As
before, detection is done by a state machine that attempts to match a known pattern of the
emulation sequence.

Removes the relevant error from client.drx-scattergather and fixes a bug in the test for AVX2
gather in 32-bit mode.

Fixes a bug in the AVX-512 state machine if index is OPSZ_8. Only OPSZ_4 is currently
tested in drx.scattergather.

Issue: #2985

### CODE REVIEW

File: ext/drx/drx.c [hunk 1]
```diff
@@ -2614,8 +2643,192 @@ static bool
...
+            reg_id_t dst0 = opnd_get_reg(instr_get_dst(&params->inst, 0));
+            reg_id_t src0 = opnd_get_reg(instr_get_src(&params->inst, 0));
+            reg_id_t src1 = opnd_get_reg(instr_get_src(&params->inst, 1));
+            if (src1 == params->the_scratch_xmm) {
+                if (src0 == params->sg_info->mask_reg) {
+                    if (dst0 == params->sg_info->mask_reg) {
+                        if (params->restore_dest_mask_start_pc <=
+                                params->info->raw_mcontext->pc &&
+                            params->info->raw_mcontext->pc <= params->prev_pc) {
+                            /* Fix the gather's destination mask here and zeroe
```
Comment (derekbruening) [inline ext/drx/drx.c]: Spelling error: "zero"

File: ext/drx/drx.c [hunk 2]
```diff
@@ -2614,8 +2643,192 @@ static bool
...
+            (params->sg_info->scalar_value_size == OPSZ_8 &&
+             instr_get_opcode(&params->inst) == OP_vpinsrq)) {
+            opnd_t src1 = instr_get_src(&params->inst, 1);
+            if (opnd_is_reg(src1)) {
+                if (opnd_get_reg(src1) == params->gpr_bit_mask) {
+                    ASSERT(opnd_is_reg(instr_get_dst(&params->inst, 0)),
+                           "internal error: unexpected instruction format");
+                    reg_id_t tmp_reg = opnd_get_reg(instr_get_dst(&params->inst, 0));
+                    if (tmp_reg == params->the_scratch_xmm) {
+                        advance_state(DRX_DETECT_RESTORE_AVX2_GATHER_EVENT_STATE_7,
```
Comment (derekbruening) [inline ext/drx/drx.c]: If advance_state() in every case increments by 1, you could just remove the 1st parameter and have it do the increment, removing a source of errors here where you have to manually get the next state right (imagining copy-paste mistakes...)?

File: ext/drx/drx.c (hunk 1 above)
Comment (hgreving2304) [reply ext/drx/drx.c]: Fixed.

File: ext/drx/drx.c (hunk 2 above)
Comment (hgreving2304) [reply ext/drx/drx.c]: I intentionally didn't do that to get the numeric meaning off the state, in case there are state transitions > 1.

**Comment by hgreving2304** [comment]
run arm tests

**Pull request merged**